packages/@react-stately/table/src/useTableColumnResizeState.ts

```diff
@@ -0,0 +1,181 @@
+    const widths = new Map<Key, number>();
+    let remainingSpace = availableSpace;
+
+    const {staticColumns, dynamicColumns} = getStaticAndDynamicColumns(affectedColumns);
+
+    staticColumns.forEach(column => {
+      let width = getRealColumnWidth(column);
+      let w = parseStaticWidth(width, tableWidth.current);
+      widths.set(column.key, w);
+      remainingSpace -= w;
+    });
+
+    // dynamic columns
+    if (dynamicColumns.length > 0) {
+      const newColumnWidths = getDynamicColumnWidths(dynamicColumns, remainingSpace, tableWidth.current);
+      for (let column of newColumnWidths) {
+        widths.set(column.key, column.calculatedWidth);
+      }
+    }
+
+    return widths;
+  }, [getStaticAndDynamicColumns, getRealColumnWidth]);
+
+  // if the columns change, need to rebuild widths.
+  useLayoutEffect(() => {
+    columnsRef.current = columns;
+    const widths = buildColumnWidths(columns, tableWidth.current);
+    setColumnWidthsForRef(widths);
+  }, [columns, buildColumnWidths]);
+
+  function setColumnWidthsForRef(newWidths: Map<Key, number>) {
+    columnWidthsRef.current = newWidths;
+    // new map so that change detection is triggered
+    setColumnWidths(newWidths);
+  }
+
+  function setTableWidth(width: number) {
+    if (width && width !== tableWidth.current) {
+      tableWidth.current = width;
+      if (!isResizing.current) {
+        const widths = buildColumnWidths(columns, width);
+        setColumnWidthsForRef(widths);
+      }
+    }
+  }
+
+  function onColumnResizeStart() {
+    isResizing.current = true;
+    startResizeContentWidth.current = getContentWidth(columnWidthsRef.current);
+  }
+
+  function onColumnResize(column: GridNode<T>, width: number) {
+    let widthsObj = resizeColumn(column, width);
+    props.onColumnResize && props.onColumnResize(widthsObj);
+  }
+
+  function onColumnResizeEnd(column: GridNode<T>, width: number) {
+    isResizing.current = false;
+    if (props.onColumnResizeEnd) {
+      let widthsObj = resizeColumn(column, width);
+      props.onColumnResizeEnd(widthsObj);
+    }
+
+    let widths = new Map<Key, number>(columnWidthsRef.current);
+    widths.set(columnsRef.current[columnsRef.current.length - 1].key, 0);
+    setColumnWidthsForRef(widths);
+  }
+
+  function resizeColumn(column: GridNode<T>, newWidth: number) : { key: Key, width: number }[] {
+    // copy the columnWidths map and set the new width for the column being resized
+    let widths = new Map<Key, number>(columnWidthsRef.current);
+    widths.set(columnsRef.current[columnsRef.current.length - 1].key, 0);
+    widths.set(column.key, Math.max(
+      getMinWidth(column.props.minWidth, tableWidth.current),
+      Math.min(Math.floor(newWidth), getMaxWidth(column.props.maxWidth, tableWidth.current))
+    ));
+
+    // keep track of all columns that have been seized
+    resizedColumnsRef.current.add(column.key);
+    setResizedColumns(resizedColumnsRef.current);
```

Comment:
this won't actually cause a re-render with state, resizedColumnsRef.current is still the same reference after `add`

do we need state for this? or is the ref enough? or do we need a new Set?

Reply:
This is a really good question and a good observation. I believe at this time we don't want a re-render and the re-render happens later. If i'm not mistaken I think we followed this pattern of using both a ref and a `useState` because that's what was happening in a lot of other places, but you're right in saying that in this case it doesn't matter.

@marshallpete for the resizedColumns, especially because the `useState` isn't actually doing anything - should we just remove this `setResizedColumns` and only use the ref?